python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+            for result in batch:
+                if isinstance(result, Exception):
+                    with self._lock:
+                        self._got_error = True
+                    if self._error_callback is not None:
+                        self._error_callback(result)
+                elif self._callback is not None:
+                    self._callback(result)
+
+            with self._lock:
+                self._num_ready += 1
+                self._results[self._indices[ready_id]] = batch
+                self._ready_index_queue.put(self._indices[ready_id])
+
+        with self._lock:
+            self._done = True
+
+    def done(self):
+        with self._lock:
+            return self._done
+
+    def got_error(self):
+        with self._lock:
+            return self._got_error
+
+    def results(self):
+        with self._lock:
+            return self._results
+
+    def next_ready_index(self, timeout=None):
+        return self._ready_index_queue.get(timeout=timeout)
+
+
+class AsyncResult(object):
+    def __init__(self,
+                 chunk_object_ids,
+                 callback=None,
+                 error_callback=None,
+                 single_result=False):
+        self._single_result = single_result
+        self._result_thread = ResultThread(chunk_object_ids, callback,
+                                           error_callback)
+        self._result_thread.start()
+
+    def wait(self, timeout=None):
+        start = time.time()
+        while timeout is None or time.time() - start < timeout:
+            if self._result_thread.done():
+                break
+            time.sleep(0.001)
+
+    def get(self, timeout=None):
+        self.wait(timeout)
+        if not self._result_thread.done():
+            raise TimeoutError
+
+        results = []
+        for batch in self._result_thread.results():
+            for result in batch:
+                if isinstance(result, PoolTaskError):
+                    raise result.underlying
+            results.extend(batch)
+        if self._single_result:
+            return results[0]
+
+        return results
+
+    def ready(self):
+        return self._result_thread.done()
+
+    def successful(self):
+        if not self.ready():
+            raise ValueError("{0!r} not ready".format(self))
+        return not self._result_thread.got_error()
+
+
+class IMapIterator(object):
+    def __init__(self, pool, func, iterable, chunksize=None):
+        self._pool = pool
+        self._func = func
+        self._next_chunk_index = 0
+        self._chunk_object_ids = []
+        self._ready_indices = []
+        self._ready_objects = []
+        if not hasattr(iterable, "__len__"):
+            iterable = [iterable]
+        self._iterator = iter(iterable)
+        if chunksize:
+            self._chunksize = chunksize
+        else:
+            self._chunksize = pool._calculate_chunksize(iterable)
+        self._total_chunks = int(math.ceil(len(iterable) / chunksize))
```

Comment:
Preferably factor this out into a function (`round_up`?) with integer-only logic, akin to the following:

    self._total_chunks = 1 + (len(iterable) - 1) // chunksize if len(iterable) > 0 else 0

Currently I have to pause and think whether the division + ceiling might have floating-point issues; it's not obvious to me. Integer-only logic is generally preferable if you can do it since it's guaranteed to have predictable behavior.